### PR TITLE
Local S3 for Development

### PR DESCRIPTION
**Pull request by richardgoater**
Key idea is the `aws` section of any config is optional and should only be present when using AWS.

### CODE REVIEW

**Comment by richardgoater** [comment]
the issue in the previous commit name was because of old config, it shouldn't affect new setups 🤞

File: metaspace/graphql/s3Client.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+export const getS3Config = () => {
+  if (config.aws) {
+    return {
+      credentials: getCredentials(config.aws),
+      region: config.aws.region,
+    }
+  }
+  const { endpoint, httpOptions, s3ForcePathStyle } = config.s3 || {}
+  return {
```
Comment (richardgoater) [inline metaspace/graphql/s3Client.ts:21]: I thought config was being spread here, I could fix it you like.

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -198,6 +198,16 @@ services:
     #  - "api:172.17.0.1"
     #  - "graphql:172.17.0.1"
 
+  storage:
+    image: minio/minio:RELEASE.2021-01-16T02-19-44Z
+    command: server /data
```
Comment (intsco) [inline docker/docker-compose.yml:203]: As the expected bucket name isn't created on startup, the very first upload fails with `NoSuchBucket: The specified bucket does not exist`.
One way of handling it here would be
```
    entrypoint: sh
    command: -c 'mkdir -p /data/sm-engine-dev && /usr/bin/minio server /data'
```

**Review by intsco — APPROVED** [review]
LGTM. I am not quite confident in my review of the JS code though.

File: metaspace/graphql/config/development.docker.js [hunk 3]
```diff
@@ -12,9 +12,31 @@ config.defaults = {
 }
 
 config.upload = {
-  destination: '/opt/data/metaspace/uploads',
+  bucket: 'sm-engine-dev',
+  moldb_prefix: 'databases',
 }
 
+config.s3 = {
+  endpoint: 'http://localhost:9000',
```
[the remark this answers is not in the document]
Comment (richardgoater) [reply metaspace/graphql/config/development.docker.js:20]: In my testing, `endpoint` needed to be the external address, one that works from the browser like `localhost`. The `httpOptions.proxy` property provides the reference to the internal `storage` host name 👍

File: docker/docker-compose.yml (hunk 2 above)
Comment (richardgoater) [reply docker/docker-compose.yml:203]: Good spot and nice idea, I'll add it.

File: metaspace/graphql/s3Client.ts (hunk 1 above)
Comment (LachlanStuart) [reply metaspace/graphql/s3Client.ts:21]: It's fine either way

File: metaspace/engine/sm/engine/molecular_db.py [hunk 4]
```diff
@@ -72,7 +74,9 @@ def _validate_moldb_df(df):
 
 def read_moldb_file(file_path):
     try:
-        moldb_df = pd.read_csv(file_path, sep='\t', dtype=object, na_filter=False)
+        bucket_name, key = split_s3_path(file_path)
+        buffer = get_s3_bucket(bucket_name).Object(key).get()['Body']
+        moldb_df = pd.read_csv(buffer, sep='\t', dtype=object, na_filter=False)
```
Comment (LachlanStuart) [inline metaspace/engine/sm/engine/molecular_db.py:79]: @intsco This no longer supports local paths, which breaks `scripts/import_molecular_db.py`.

**Review by LachlanStuart — APPROVED** [review]
I gave it a try and it works great. 

There's one issue with `read_moldb_file` not supporting local files anymore, which should be fixed. However, it's fine if that comes in a follow-up PR.

**Comment by intsco** [comment]
@richardgoater When do you plan to merge this PR? I want to start working on my next task based on the changes in this PR.

**Comment by richardgoater** [comment]
@intsco I can merge it now

**Pull request merged**